### PR TITLE
Speedup scoring

### PR DESCRIPTION
**Pull request by gnusi**

### CODE REVIEW

File: core/utils/misc.hpp [hunk 1]
```diff
@@ -107,40 +108,44 @@ move_on_copy<T> make_move_on_copy(T&& value) noexcept {
...
 template<
-  typename Input,
-  Input Size,
-  typename Func,
-  typename = typename std::enable_if_t<std::is_integral_v<Input>>
-> class cached_func {
+    typename Input,
+    Input Size,
+    typename Func,
+    typename = typename std::enable_if_t<std::is_integral_v<Input>>>
```
Comment (MBkkt) [inline core/utils/misc.hpp:114]: When signed?

File: tests/utils/misc_test.cpp [hunk 2]
```diff
@@ -27,11 +27,11 @@ TEST(misc_test, cached_func_contexpr) {
   constexpr auto cached_func = irs::cache_func<uint32_t, 3>(
     0, [](uint32_t v) { return v + 1; });
   static_assert (3 == cached_func.size());
-  static_assert (1 == cached_func(0));
-  static_assert (2 == cached_func(1));
-  static_assert (3 == cached_func(2));
-  static_assert (4 == cached_func(3));
-  static_assert (5 == cached_func(4));
+  static_assert (1 == cached_func.get<false>(0));
```
Comment (MBkkt) [inline tests/utils/misc_test.cpp:30]: Why false?

File: core/utils/misc.hpp [hunk 3]
```diff
@@ -24,6 +24,7 @@
 #define IRESEARCH_MISC_H
 
 #include <array>
+#include <assert.h>
```
Comment (MBkkt) [inline core/utils/misc.hpp:27]:
```suggestion
#include <cassert>
```

File: core/utils/misc.hpp [hunk 4]
```diff
@@ -107,40 +108,44 @@ move_on_copy<T> make_move_on_copy(T&& value) noexcept {
...
-    return IRESEARCH_COUNTOF(cache_);
+    return cache_.size();
   }
 
  private:
   Func func_;
-  output_type cache_[Size]{};
-
-  static_assert(IRESEARCH_COUNTOF(decltype(cache_){}) == Size);
+  std::array<output_type, Size> cache_;
```
Comment (MBkkt) [inline core/utils/misc.hpp:144]: The old cache was initialized, now not, is it ok?

File: core/search/tfidf.cpp [hunk 5]
```diff
@@ -264,35 +264,46 @@ struct idf final {
...
       const tfidf::idf& idf,
       const frequency* freq,
-      const filter_boost* fb = nullptr) noexcept
-    : score_buf(score_buf),
-      freq(freq ? freq : &EMPTY_FREQ),
-      filter_boost(fb),
-      idf(boost * idf.value) {
+      const irs::filter_boost* filter_boost = nullptr) noexcept
+    : score_buf{score_buf},
+      freq{freq ? freq : &EMPTY_FREQ},
```
Comment (MBkkt) [inline core/search/tfidf.cpp:275]: Can we make a ref instead of a pointer for freq?

File: core/search/bm25.cpp [hunk 6]
```diff
@@ -300,80 +300,187 @@ namespace bm25 {
...
+            assert(state.filter_boost);
+            c0 = state.filter_boost->value * state.num;
+          } else {
+            c0 = state.num;
+          }
+
+          float_t c1;
+          if constexpr (NormType::kNorm2Tiny == Norm::kType) {
+            static_assert(std::is_same_v<uint32_t, decltype(state.norm())>);
+            c1 = state.norm_cache[state.norm() & 0xFF];
```
Comment (MBkkt) [inline core/search/bm25.cpp:462]: NIT
```suggestion
            c1 = state.norm_cache[state.norm() & 0xFFU];
```

File: core/search/bm25.cpp [hunk 7]
```diff
@@ -300,80 +300,187 @@ namespace bm25 {
...
+          } else {
+            c1 = 1.f/(state.norm_const + state.norm_length * state.norm());
```
Comment (MBkkt) [inline core/search/bm25.cpp:464]: Why can't we invert value in cache and replace `tf * c1` to `tf / c1` and then remove `1.f / `
Also, I think this
```C++
...(state.score_buf) = c0 - c0 / (1.f + tf / c1);
```
can be replaced by something like this
```C++
...(state.score_buf) = c0 - c0 * c1 / (c1 + tf);
```

File: core/search/bm25.cpp [hunk 8]
```diff
@@ -300,80 +300,187 @@ namespace bm25 {
...
-  float_t norm_const; // precomputed k*(1-b)
-  float_t norm_length; // precomputed k*b/avgD
+  // precomputed idf value
+  float_t idf;
+  // precomputed k*(1-b)
+  float_t norm_const;
+  // precomputed k*b/avgD
+  float_t norm_length;
+  // precomputed 1/(k*(1-b+b*|doc|/avgDL)) for |doc| E [0..255]
+  float_t norm_cache[256];
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/search/bm25.cpp:311]: `stats` it must be POD

File: core/search/bm25.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (gnusi) [reply core/search/bm25.cpp:462]: Done

File: core/search/bm25.cpp [hunk 9]
```diff
@@ -418,10 +525,16 @@ class sort final : public irs::prepared_sort_basic<bm25::score_t, bm25::stats> {
...
+      const float_t avg_dl = float_t(total_term_freq) / docs_with_field;
+      stats.norm_length = kb / avg_dl;
+    } else {
+      stats.norm_length = kb;
+    }
+
+    auto begin = std::begin(stats.norm_cache);
+    for (uint32_t i = 0; i < IRESEARCH_COUNTOF(stats.norm_cache); ++i) {
+      *begin++ = 1.f/(stats.norm_const + stats.norm_length * i);
     }
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/search/bm25.cpp:539]: Done

File: core/utils/misc.hpp (hunk 4 above)
Comment (gnusi) [reply core/utils/misc.hpp:144]: Good catch.

File: core/utils/misc.hpp (hunk 3 above)
Comment (gnusi) [reply core/utils/misc.hpp:27]: Done

File: core/utils/misc.hpp [hunk 10]
```diff
@@ -107,40 +108,44 @@ move_on_copy<T> make_move_on_copy(T&& value) noexcept {
...
   using input_type = Input;
   using output_type = std::invoke_result_t<Func, Input>;
 
-  constexpr explicit cached_func(size_t offset, Func&& func)
+  constexpr explicit cached_func(input_type offset, Func&& func)
     : func_{std::forward<Func>(func)} {
-    for (input_type i = offset; i < size(); ++i) {
+    for (input_type i{offset}; i < size(); ++i) {
       cache_[i] = func_(i);
     }
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/utils/misc.hpp:124]: Done

File: tests/utils/misc_test.cpp (hunk 2 above)
Comment (gnusi) [reply tests/utils/misc_test.cpp:30]: to check "Unchecked" case

File: core/search/tfidf.cpp [hunk 11]
```diff
@@ -264,35 +264,46 @@ struct idf final {
...
   }
 
   FORCE_INLINE float_t operator()() {
-    if constexpr (IsNorm2) {
-      return RSQRT(reader());
+    if constexpr (kType < NormType::kNorm) {
+      return kRSQRT.get<kType != NormType::kNorm2Tiny>(reader());
     }
 
     return reader();
```
[the remark this answers is not in the document]
Comment (gnusi) [reply core/search/tfidf.cpp:309]: Done

File: core/utils/misc.hpp (hunk 1 above)
Comment (gnusi) [reply core/utils/misc.hpp:114]: Do you see any issues with signed/unsigned types?

Comment (MBkkt) [reply core/utils/misc.hpp:114]: Not, you right, I thought if unsigned we can replace `<` to `!=`, but this is not true

**Review by MBkkt — APPROVED** [review]

**Review by Dronplane — APPROVED** [review]
LGTM

**Pull request merged**